ext/libxml/ruby_xml_node.c

```diff
@@ -45,13 +45,18 @@ static void rxml_node_deregisterNode(xmlNodePtr xnode)
   /* Has the node been wrapped and exposed to Ruby? */
   if (xnode->_private)
   {
+    /* Does it belong to another libxml user? */
+    if (!OWNED(xnode))
```

Comment:
That's a cleaner solution, but it turns it into a performance question again: now you have to look up every node unconditionally on destruction, even if it's owned by another library (be it a gem or an independent C library). I don't actually know what proportion of nodes have their `_private` set in the general case, but it seems a simple enough optimisation to keep.